onnx/optimizer/pass.h

```diff
@@ -0,0 +1,195 @@
+  }
+  virtual std::string getPassName() const = 0;
+
+  virtual bool initializePass(Graph& graph) {
+    return false;
+  }
+  virtual bool finalizePass(Graph& graph) {
+    return false;
+  }
+  virtual std::shared_ptr<PostPassAnalysis> runPass(Graph& graph) = 0;
+
+ protected:
+  // Iterates through the elements in the graph and counts the number of times
+  // the transform is succesfully run.
+  unsigned int DescendOnGraphAttributesAndCount(
+      Node* n,
+      std::function<unsigned int(Graph&)> fn);
+  // A more general version of the function above that doesn't constrain the
+  // return type of fn.
+  void DescendOnGraphAttributesUnconstrained(
+      Node* n,
+      std::function<void(Graph&)> fn);
+};
+
+class ImmutablePass : Pass {
+ public:
+  explicit ImmutablePass()
+      : Pass(
+            PassType::Immutable,
+            PassEfficiency::Complete,
+            PassOptimizationType::None) {}
+  ~ImmutablePass() override;
+};
+
+// Pass Analysis done after a predicate based pass.
+struct CountBasedPassAnalysis : PostPassAnalysis {
+  // Have to use raw pointer here. The idea is that the pass will pass <this> as
+  // a parameter to the constructor. We could use std::enable_shared_from_this
+  // but this complicates the memory model. Also since all passes come from
+  // GlobalPassRegistry which already utilizes smart pointers we don't have to
+  // worry about memory leaks from passes.
+  Pass* pass;
+  unsigned int num_positive_transforms;
+  bool initialization_done;
+  bool finalization_done;
+
+ public:
+  explicit CountBasedPassAnalysis(
+      Pass* pass,
+      unsigned int num_positive_transforms,
+      bool initialization_done,
+      bool finalization_done);
+
+  bool graphChanged() {
+    return this->num_positive_transforms > 0;
+  }
+  bool numSucceededTransforms() {
+    return this->num_positive_transforms;
+  }
+
+  // Whether or not a repeated application of the pass might be useful.
+  bool fixedPointOptimizationNeeded() {
+    return this->graphChanged() &&
+        pass->getPassEfficiency() == PassEfficiency::Partial;
+  }
+};
+
+// A pass that is based on pattern matching. The majority of passes will
+// implement this pass. In order for the pass to work the patternMatchPredicate
+// function must be implemented matches a subgraph to the respective
+// optimization pass. Lastly the runTransform method must also be implemented
+// which simply implements the pass on any node which passes
+// patternMatchPredicate.
+class PredicateBasedPass : public Pass {
+ public:
+  explicit PredicateBasedPass(
+      PassType pass_type,
+      PassEfficiency pass_efficiency,
+      PassOptimizationType pass_optimization_type)
+      : Pass(pass_type, pass_efficiency, pass_optimization_type) {}
+  ~PredicateBasedPass() override;
+
+  virtual bool patternMatchPredicate(Node* node) = 0;
+  virtual bool
+  runTransform(Node* node, Graph& graph, NodeDestroyType& destroy_current) = 0;
```

Comment:
Add some comments about what should be put in the `runTransform`? (do everything but not deleting the node?)